src/plots/histogram.rs

```diff
@@ -0,0 +1,440 @@
+    ///
+    /// Histogram::new(0, f64::NAN, f64::INFINITY, Measurement::Unitless).unwrap_err();
+    /// ```
+    pub fn new(
+        number_of_buckets: usize,
+        min: f64,
+        max: f64,
+        measurement: Measurement,
+    ) -> Result<Self> {
+        ensure!(
+            number_of_buckets > 0,
+            error::PlotError {
+                details: "Histograms must have at least one bucket"
+            }
+        );
+        ensure!(
+            min.is_finite() && max.is_finite(),
+            error::PlotError {
+                details: "Histograms must have finite min/max values"
+            }
+        );
+        ensure!(
+            min < max || (approx_eq!(f64, min, max) && number_of_buckets == 1),
+            error::PlotError {
+                details: "Histograms max value must be larger than its min value"
+            }
+        );
+
+        Ok(Self {
+            counts: vec![0; number_of_buckets],
+            labels: None,
+            nodata_count: 0,
+            min,
+            max,
+            measurement,
+        })
+    }
+
+    /// Creates a new empty histogram
+    ///
+    /// # Examples
+    /// ```rust
+    /// use geoengine_datatypes::plots::Histogram;
+    /// use geoengine_datatypes::primitives::Measurement;
+    /// use std::f64;
+    ///
+    /// Histogram::new_with_labels(2, 0., 1., Measurement::Unitless, vec!["foo".into(), "bar".into()]).unwrap();
+    ///
+    /// Histogram::new_with_labels(0, f64::NAN, f64::INFINITY, Measurement::Unitless, vec!["foo".into(), "bar".into()]).unwrap_err();
+    /// Histogram::new_with_labels(2, 0., 1., Measurement::Unitless, vec!["foo".into()]).unwrap_err();
+    /// ```
+    pub fn new_with_labels(
+        number_of_buckets: usize,
+        min: f64,
+        max: f64,
+        measurement: Measurement,
+        labels: Vec<String>,
+    ) -> Result<Self> {
+        ensure!(
+            number_of_buckets > 0,
+            error::PlotError {
+                details: "Histograms must have at least one bucket"
+            }
+        );
+        ensure!(
+            min.is_finite() && max.is_finite(),
+            error::PlotError {
+                details: "Histograms must have finite min/max values"
+            }
+        );
+        ensure!(
+            min < max || (approx_eq!(f64, min, max) && number_of_buckets == 1),
+            error::PlotError {
+                details: "Histograms max value must be larger than its min value"
+            }
+        );
+        ensure!(
+            labels.len() == number_of_buckets,
+            error::PlotError {
+                details: "Histogram must have as many labels as buckets"
+            }
+        );
+
+        Ok(Self {
+            counts: vec![0; number_of_buckets],
+            labels: Some(labels),
+            nodata_count: 0,
+            min,
+            max,
+            measurement,
+        })
+    }
+
+    pub fn add_feature_data(&mut self, data: FeatureDataRef) -> Result<()> {
```

Comment:
there is a lot of code duplication. maybe implement an iterator on FeatureDataRef that yields numeric values? Could also be implemented for rasters because it's always the same how to generate the histogram

Reply:
Generally, yes, but I'm not quite sure if this would slow down OpenCL implementations.
Maybe I could cast the underlying Arrow array to have at least SIMD.

Reply:
how does this code relate to future opencl implementations?

Reply:
I would rather not encourage value-by-value iterators when we aim for parallel implementations with OpenCL.